patches/api/0421-Add-damage-source-wrapper.patch

```diff
@@ -0,0 +1,2335 @@
++        return ofType(VanillaDamageType.PLAYER_ATTACK, damager);
++    }
++
++    /**
++     * Generic arrow damage origin is dealt when an arrow hits
++     * a target.
++     * Trident should use the {@link #ofTrident(Trident)} instead.
++     *
++     * @param arrow The arrow
++     * @return      The damage origin created
++     * @see VanillaDamageType#ARROW
++     */
++    @NotNull
++    public static DamageOrigin ofArrow(@NotNull AbstractArrow arrow) {
++        return ofArrow(arrow, null);
++    }
++
++    /**
++     * Generic arrow damage origin is dealt when an arrow hits
++     * a target.
++     * Trident should use the {@link #ofTrident(Trident, Entity)} instead.
++     *
++     * @param arrow   The arrow
++     * @param thrower The thrower
++     * @return        The damage origin created
++     * @see VanillaDamageType#ARROW
++     */
++    @NotNull
++    public static DamageOrigin ofArrow(@NotNull AbstractArrow arrow, @Nullable Entity thrower) {
++        if (arrow instanceof Trident) { // Thanks bukkit to consider trident as an arrow o_o
++            throw new IllegalArgumentException("Projectile can only be a variant of arrow. Trident has its own method.");
++        }
++        return ofType(VanillaDamageType.ARROW, arrow, thrower);
++    }
++
++    /**
++     * Trident damage origin is dealt when a trident hits a target.
++     * @param trident The trident
++     * @return        The damage origin created
++     * @see VanillaDamageType#TRIDENT
++     */
++    @NotNull
++    public static DamageOrigin ofTrident(@NotNull Trident trident) {
++        return ofTrident(trident, null);
++    }
++
++    /**
++     * Trident damage origin is dealt when a trident hits a target.
++     * @param projectile The projectile
++     * @param thrower    The thrower
++     * @return           The damage origin created
++     * @see VanillaDamageType#TRIDENT
++     */
++    @NotNull
++    public static DamageOrigin ofTrident(@NotNull Trident projectile, @Nullable Entity thrower) {
++        return ofType(VanillaDamageType.TRIDENT, projectile, thrower);
++    }
++
++    /**
++     * Mob projectile damage origin is dealt when a projectile launched by a mob
++     * hit a target.
++     *
++     * @param projectile The projectile
++     * @return           The damage origin created
++     * @see VanillaDamageType#MOB_PROJECTILE
++     */
++    @NotNull
++    public static DamageOrigin ofMobProjectile(@NotNull Entity projectile) {
++        return ofMobProjectile(projectile, null);
++    }
++
++    /**
++     * Mob projectile damage origin is dealt when a projectile launched by a mob
++     * hit a target.
++     *
++     * @param projectile The projectile
++     * @param thrower    The thrower
++     * @return           The damage origin created
++     * @see VanillaDamageType#MOB_PROJECTILE
++     */
++    @NotNull
++    public static DamageOrigin ofMobProjectile(@NotNull Entity projectile, @Nullable Entity thrower) {
```

Comment:
Should the thrower be a mob?

Reply:
In vanilla they only allow living entity. i think i will migrate to that instead